File: sdk/java-client/src/test/java/dev/latent/sdk/ProfileLifetime.java
package dev.latent.sdk;

import java.nio.ByteBuffer;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;

final class ProfileLifetime {
    private ProfileLifetime() { }

    private static void check(boolean condition, String message) {
        if (!condition) throw new AssertionError(message);
    }

    private static Management.ResponseMetadata metadata(String operationId, Management.OutcomeKnowledge outcome) {
        return new Management.ResponseMetadata(new Management.RequestIdentity(Optional.empty(), Optional.ofNullable(operationId)), outcome, Optional.empty(), Optional.empty());
    }

    private static <Response> CompletableFuture<Management.ClientResponse<Response>> ready(Response value) {
        return CompletableFuture.completedFuture(new Management.ClientResponse<>(value, metadata(null, Management.OutcomeKnowledge.OBSERVED)));
    }

    private static Management.ClientFailure failure(String operationId, boolean dispatched, Management.FailureCategory category) {
        return new Management.ClientFailure(category, "local-fixture-failure", Optional.empty(), Optional.empty(), dispatched,
                dispatched ? Management.OutcomeKnowledge.UNKNOWN : Management.OutcomeKnowledge.NOT_DISPATCHED,
                new Management.RequestIdentity(Optional.empty(), Optional.of(operationId)), Optional.empty(), Optional.empty(), Optional.empty());
    }

    private static final class FixtureClient implements Management.ClientProfile {
        private int writes;
        private int waiters;
        private int cancels;
        private int pages;
        private Management.CapabilityPolicyOperation receipt;
        private Management.Policy policy;

        @Override
        public CompletableFuture<Management.ClientResponse<Management.InvokeResponse>> invoke(Management.InvokeRequest request, Management.CallOptions options) {
            byte[] owned = new byte[request.payload().remaining()];
            request.payload().duplicate().get(owned);
            var success = new Management.Success(ByteBuffer.wrap(owned), request.mediaType(), Optional.empty(), List.of(), Map.of());
            return ready(new Management.InvokeResponse(request.activationId().orElseThrow(), "revision-a", "component-a", 1,
                    Optional.of(success), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()));
        }

        @Override
        public CompletableFuture<Management.ClientResponse<Management.CancelResponse>> cancel(Management.CancelRequest request, Management.CallOptions options) {
            cancels++;
            return ready(new Management.CancelResponse(Management.CancelDisposition.ACCEPTED, Optional.empty()));
        }

        @Override
        public CompletableFuture<Management.ClientResponse<Management.ActivationStatus>> getActivation(Management.GetActivationRequest request, Management.CallOptions options) {
            return ready(new Management.ActivationStatus(request.activationId(), "running", Optional.empty(), 0, Map.of(),
                    Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty(), Optional.empty()));
        }

        @Override
        public CompletableFuture<Management.ClientResponse<Management.GetPolicyResponse>> getPolicy(Management.GetPolicyRequest request, Management.CallOptions options) {
            return ready(new Management.GetPolicyResponse(Optional.ofNullable(policy)));
        }

        @Override
        public CompletableFuture<Management.ClientResponse<Management.ListPoliciesResponse>> listPolicies(Management.ListPoliciesRequest request, Management.CallOptions options) {
            pages++;
            return ready(new Management.ListPoliciesResponse(List.of(), 1, Optional.of(new Management.PageResponse(Optional.of("opaque-next-page")))));
        }

        @Override
        public CompletableFuture<Management.ClientResponse<Management.ListCapabilitiesResponse>> listCapabilities(Management.ListCapabilitiesRequest request, Management.CallOptions options) {
            var revision = new Management.CapabilityInspectionRevision(request.deploymentId(), "revision-a", "component-a", Optional.empty(), 1, 1);
            return ready(new Management.ListCapabilitiesResponse(List.of(), Optional.empty(), Optional.of(revision), Optional.empty(), Optional.empty(), "binding-plan-unavailable"));
        }

        @Override
        public CompletableFuture<Management.ClientResponse<Management.ApplyPolicyResponse>> applyPolicy(Management.ApplyPolicyRequest request, Management.CallOptions options) {
            if (options.timeoutMillis().isPresent() && options.timeoutMillis().get() == 0) {
                return CompletableFuture.failedFuture(new Management.ClientException(failure(request.operationId(), false, Management.FailureCategory.DEADLINE)));
            }
            check(request.expectedGeneration().isPresent() && !request.operationId().isEmpty(), "explicit mutation identity and precondition");
            if (receipt != null) return ready(new Management.ApplyPolicyResponse(Optional.empty(), Optional.of(receipt)));
            policy = request.policy().orElseThrow();
            receipt = new Management.CapabilityPolicyOperation(request.operationId(), "tenant-a", policy.id(), policy.recordKind(), -1L, "digest-a", false);
            writes++;
            waiters++;
            return new CompletableFuture<>() {
                @Override
                public boolean cancel(boolean mayInterruptIfRunning) {
                    boolean changed = completeExceptionally(new Management.ClientCancellationException(failure(request.operationId(), true, Management.FailureCategory.LOCAL_CANCELLED)));
                    if (changed) waiters--;
                    return changed;
                }
            };
        }

        @Override
        public CompletableFuture<Management.ClientResponse<Management.GetPolicyOperationResponse>> getPolicyOperation(Management.GetPolicyOperationRequest request, Management.CallOptions options) {
            var found = Optional.ofNullable(receipt).filter(value -> value.operationId().equals(request.operationId()));
            return CompletableFuture.completedFuture(new Management.ClientResponse<>(new Management.GetPolicyOperationResponse(found),
                    metadata(request.operationId(), found.isPresent() ? Management.OutcomeKnowledge.OBSERVED : Management.OutcomeKnowledge.UNKNOWN)));
        }
    }

    static void run() throws Exception {
        var client = new FixtureClient();
        Management.ClientProfile profile = client;
        var options = new Management.CallOptions(Optional.empty());
        var policy = new Management.Policy("policy-a", Optional.empty(), "{}", 0, "lsf-capability-policy-v1", Management.CapabilityPolicyRecordKind.POLICY, "", false);
        var request = new Management.ApplyPolicyRequest(Optional.of(policy), Optional.of(0L), "operation-a");
        var pending = profile.applyPolicy(request, options);
        check(client.writes == 1 && client.waiters == 1, "server receipt precedes local completion");
        check(pending.cancel(false) && pending.isCancelled(), "local future cancellation");
        try {
            pending.join();
            throw new AssertionError("cancelled wait must fail");
        } catch (java.util.concurrent.CancellationException cancelled) {
            var retained = Management.clientFailure(cancelled).orElseThrow();
            check(retained.dispatched() && retained.outcome().equals(Management.OutcomeKnowledge.UNKNOWN), "local cancellation is not rollback");
            check(retained.identity().operationId().orElseThrow().equals("operation-a"), "original recovery identity");
        }
        check(client.waiters == 0 && client.writes == 1 && client.cancels == 0, "only local ownership ends");
        var recovered = profile.getPolicyOperation(new Management.GetPolicyOperationRequest("operation-a"), options).get(2, TimeUnit.SECONDS);
        check(recovered.value().receipt().orElseThrow().generation() == -1L && recovered.metadata().auditAck().isEmpty(), "u64 receipt and no invented audit");
        var unknown = profile.getPolicyOperation(new Management.GetPolicyOperationRequest("not-retained"), options).get(2, TimeUnit.SECONDS);
        check(unknown.value().receipt().isEmpty() && unknown.metadata().outcome().equals(Management.OutcomeKnowledge.UNKNOWN), "missing receipt is unknown");
        profile.applyPolicy(request, options).get(2, TimeUnit.SECONDS);
        check(client.writes == 1, "explicit replay does not mutate again");
        try {
            profile.applyPolicy(request, new Management.CallOptions(Optional.of(0L))).get(2, TimeUnit.SECONDS);
            throw new AssertionError("zero timeout must fail before dispatch");
        } catch (ExecutionException expired) {
            check(expired.getCause() instanceof Management.ClientException failure && !failure.failure().dispatched(), "typed zero-timeout failure");
        }
        byte[] payload = new byte[]{1, 2};
        var invocation = new Management.InvokeRequest(Optional.of("activation-a"), Optional.empty(), Optional.empty(), Optional.empty(), ByteBuffer.wrap(payload), "application/octet-stream", Optional.empty(), 0, Optional.empty(), Optional.empty(), Map.of());
        var invoked = profile.invoke(invocation, options).get(2, TimeUnit.SECONDS);
        payload[0] = 99;
        check(invoked.value().success().orElseThrow().payload().get(0) == 1, "response owns bytes");
        check(profile.cancel(new Management.CancelRequest("activation-a", "fixture"), options).get(2, TimeUnit.SECONDS).value().disposition().equals(Management.CancelDisposition.ACCEPTED), "explicit remote Cancel");
        check(profile.getActivation(new Management.GetActivationRequest("activation-a"), options).get(2, TimeUnit.SECONDS).value().phase().equals("running"), "accepted does not prove cleanup");
        check(profile.getPolicy(new Management.GetPolicyRequest("policy-a", Management.CapabilityPolicyRecordKind.POLICY), options).get(2, TimeUnit.SECONDS).value().policy().isPresent(), "policy inspection");
        var page = profile.listPolicies(new Management.ListPoliciesRequest(Management.CapabilityPolicyRecordKind.POLICY, Optional.of(new Management.PageRequest(1, Optional.empty()))), options).get(2, TimeUnit.SECONDS);
        check(page.value().page().orElseThrow().nextPageToken().isPresent() && client.pages == 1, "one bounded page without draining");
        var capabilities = profile.listCapabilities(new Management.ListCapabilitiesRequest(Optional.empty(), Optional.empty(), Optional.empty(), "deployment-a", false), options).get(2, TimeUnit.SECONDS);
        check(capabilities.value().revision().orElseThrow().deploymentId().equals("deployment-a"), "selected deployment");
        System.out.println("shared profile lifetime/recovery: passed");
    }
}

File: sdk/dotnet/Latent.Sdk.SemanticTests/ProfileLifetime.cs
using Profile = Latent.Sdk.Profile;

namespace Latent.Sdk.SemanticTests;

internal static class ProfileLifetime
{
    private static void Check(bool condition, string message)
    {
        if (!condition) throw new InvalidOperationException(message);
    }

    private static Profile.ResponseMetadata Metadata(string? operationId, Profile.OutcomeKnowledge outcome) =>
        new(new Profile.RequestIdentity(null, operationId), outcome, null, null, null);

    private static Profile.ClientFailure Failure(string operationId, bool dispatched, Profile.FailureCategory category) =>
        new(category, "local-fixture-failure", null, null, dispatched,
            dispatched ? Profile.OutcomeKnowledge.Unknown : Profile.OutcomeKnowledge.NotDispatched,
            new Profile.RequestIdentity(null, operationId), null, null, null, null);

    private static ValueTask<Profile.ClientResponse<Response>> Reply<Response>(Response value) =>
        ValueTask.FromResult(new Profile.ClientResponse<Response>(value, Metadata(null, Profile.OutcomeKnowledge.Observed)));

    private sealed class FixtureClient : Profile.IClientProfile
    {
        internal int Writes;
        internal int Waiters;
        internal int Cancels;
        internal int Pages;
        private Profile.CapabilityPolicyOperation? receipt;
        private Profile.Policy? policy;

        public ValueTask<Profile.ClientResponse<Profile.InvokeResponse>> InvokeAsync(Profile.InvokeRequest request, Profile.CallOptions options, CancellationToken cancellationToken = default) =>
            Reply(new Profile.InvokeResponse(request.ActivationId!, "revision-a", "component-a", 1,
                new Profile.Success(request.Payload.ToArray(), request.MediaType, null, Array.Empty<string>(), new Dictionary<string, string>()), null, null, null, null));

        public ValueTask<Profile.ClientResponse<Profile.CancelResponse>> CancelAsync(Profile.CancelRequest request, Profile.CallOptions options, CancellationToken cancellationToken = default)
        {
            Cancels++;
            return Reply(new Profile.CancelResponse(Profile.CancelDisposition.Accepted, null));
        }

        public ValueTask<Profile.ClientResponse<Profile.ActivationStatus>> GetActivationAsync(Profile.GetActivationRequest request, Profile.CallOptions options, CancellationToken cancellationToken = default) =>
            Reply(new Profile.ActivationStatus(request.ActivationId, "running", null, 0, new Dictionary<string, string>(), null, null, null, null, null));

        public ValueTask<Profile.ClientResponse<Profile.GetPolicyResponse>> GetPolicyAsync(Profile.GetPolicyRequest request, Profile.CallOptions options, CancellationToken cancellationToken = default) =>
            Reply(new Profile.GetPolicyResponse(policy));

        public ValueTask<Profile.ClientResponse<Profile.ListPoliciesResponse>> ListPoliciesAsync(Profile.ListPoliciesRequest request, Profile.CallOptions options, CancellationToken cancellationToken = default)
        {
            Pages++;
            return Reply(new Profile.ListPoliciesResponse(Array.Empty<Profile.Policy>(), 1, new Profile.PageResponse("opaque-next-page")));
        }

        public ValueTask<Profile.ClientResponse<Profile.ListCapabilitiesResponse>> ListCapabilitiesAsync(Profile.ListCapabilitiesRequest request, Profile.CallOptions options, CancellationToken cancellationToken = default) =>
            Reply(new Profile.ListCapabilitiesResponse(Array.Empty<Profile.CapabilityDescriptor>(), null,
                new Profile.CapabilityInspectionRevision(request.DeploymentId, "revision-a", "component-a", null, 1, 1), null, null, "binding-plan-unavailable"));

        public async ValueTask<Profile.ClientResponse<Profile.ApplyPolicyResponse>> ApplyPolicyAsync(Profile.ApplyPolicyRequest request, Profile.CallOptions options, CancellationToken cancellationToken = default)
        {
            if (cancellationToken.IsCancellationRequested)
                throw new Profile.ClientCancellationException(Failure(request.OperationId, false, Profile.FailureCategory.LocalCancelled), cancellationToken);
            if (options.TimeoutMillis == 0)
                throw new Profile.ClientException(Failure(request.OperationId, false, Profile.FailureCategory.Deadline));
            Check(request.ExpectedGeneration.HasValue && request.OperationId != "" && request.Policy is not null, "explicit mutation precondition");
            if (receipt is not null)
                return new(new Profile.ApplyPolicyResponse(null, receipt), Metadata(request.OperationId, Profile.OutcomeKnowledge.Observed));
            policy = request.Policy!;
            receipt = new Profile.CapabilityPolicyOperation(request.OperationId, "tenant-a", policy.Id, policy.RecordKind, ulong.MaxValue, "digest-a", false);
            Writes++;
            Waiters++;
            try
            {
                await Task.Delay(Timeout.Infinite, cancellationToken);
                throw new InvalidOperationException("fixture requires local cancellation");
            }
            catch (OperationCanceledException)
            {
                throw new Profile.ClientCancellationException(Failure(request.OperationId, true, Profile.FailureCategory.LocalCancelled), cancellationToken);
            }
            finally
            {
                Waiters--;
            }
        }

        public ValueTask<Profile.ClientResponse<Profile.GetPolicyOperationResponse>> GetPolicyOperationAsync(Profile.GetPolicyOperationRequest request, Profile.CallOptions options, CancellationToken cancellationToken = default)
        {
            var found = receipt?.OperationId == request.OperationId ? receipt : null;
            return ValueTask.FromResult(new Profile.ClientResponse<Profile.GetPolicyOperationResponse>(new(found),
                Metadata(request.OperationId, found is null ? Profile.OutcomeKnowledge.Unknown : Profile.OutcomeKnowledge.Observed)));
        }
    }

    internal static async Task Run()
    {
        var client = new FixtureClient();
        Profile.IClientProfile profile = client;
        var options = new Profile.CallOptions(null);
        var policy = new Profile.Policy("policy-a", null, "{}", 0, "lsf-capability-policy-v1", Profile.CapabilityPolicyRecordKind.Policy, "", false);
        var request = new Profile.ApplyPolicyRequest(policy, 0, "operation-a");
        using var cancellation = new CancellationTokenSource();
        var pending = profile.ApplyPolicyAsync(request, options, cancellation.Token);
        Check(client.Writes == 1 && client.Waiters == 1, "server receipt precedes local completion");
        cancellation.Cancel();
        try
        {
            await pending.AsTask().WaitAsync(TimeSpan.FromSeconds(2));
            throw new InvalidOperationException("cancelled wait must fail");
        }
        catch (Profile.ClientCancellationException cancelled)
        {
            Check(cancelled.Failure.Dispatched && cancelled.Failure.Outcome == Profile.OutcomeKnowledge.Unknown, "local cancellation is not rollback");
            Check(cancelled.Failure.Identity.OperationId == "operation-a", "recovery identity survives");
        }
        Check(client.Waiters == 0 && client.Writes == 1 && client.Cancels == 0, "only local ownership ends");
        var recovered = await profile.GetPolicyOperationAsync(new("operation-a"), options);
        Check(recovered.Value.Receipt?.Generation == ulong.MaxValue && recovered.Metadata.AuditAck is null, "original receipt without invented audit");
        var unknown = await profile.GetPolicyOperationAsync(new("not-retained"), options);
        Check(unknown.Value.Receipt is null && unknown.Metadata.Outcome == Profile.OutcomeKnowledge.Unknown, "missing receipt remains unknown");
        await profile.ApplyPolicyAsync(request, options);
        Check(client.Writes == 1, "explicit replay does not mutate again");
        try
        {
            await profile.ApplyPolicyAsync(request, new(0));
            throw new InvalidOperationException("zero timeout must fail before dispatch");
        }
        catch (Profile.ClientException expired)
        {
            Check(!expired.Failure.Dispatched, "zero timeout stays local");
        }
        byte[] payload = [1, 2];
        var invocation = new Profile.InvokeRequest("activation-a", null, null, null, payload, "application/octet-stream", null, 0, null, null, new Dictionary<string, string>());
        var invoked = await profile.InvokeAsync(invocation, options);
        payload[0] = 99;
        Check(invoked.Value.Success!.Payload.Span[0] == 1, "returned bytes are owned");
        Check((await profile.CancelAsync(new("activation-a", "fixture"), options)).Value.Disposition == Profile.CancelDisposition.Accepted, "explicit remote Cancel");
        Check((await profile.GetActivationAsync(new("activation-a"), options)).Value.Phase == "running", "accepted does not imply cleanup");
        Check((await profile.GetPolicyAsync(new("policy-a", Profile.CapabilityPolicyRecordKind.Policy), options)).Value.Policy?.Id == "policy-a", "policy inspection");
        Check((await profile.ListPoliciesAsync(new(Profile.CapabilityPolicyRecordKind.Policy, new(1, null)), options)).Value.Page?.NextPageToken is not null && client.Pages == 1, "one bounded page without draining");
        Check((await profile.ListCapabilitiesAsync(new(null, null, null, "deployment-a", false), options)).Value.Revision?.DeploymentId == "deployment-a", "explicit deployment");
        Console.WriteLine("shared profile lifetime/recovery: passed");
    }
}
